src/Features/LanguageServer/Protocol/Handler/Diagnostics/Public/PublicWorkspacePullDiagnosticsHandler.cs

```diff
@@ -118,4 +137,55 @@ protected override WorkspaceDiagnosticPartialReport CreateUnchangedReport(TextDo
             }
         }).ToImmutableArray();
     }
+
+    private void OnLspSolutionChanged(object? sender, WorkspaceChangeEventArgs e)
+    {
+        UpdateLspChanged();
+    }
+
+    private void OnLspTextChanged(object? sender, EventArgs e)
+    {
+        UpdateLspChanged();
+    }
+
+    private void UpdateLspChanged()
+    {
+        Interlocked.Exchange(ref _lspChanged, 1);
+    }
+
+    protected override async Task WaitForChangesAsync(RequestContext context, CancellationToken cancellationToken)
+    {
+        // Spin waiting until our LSP change flag has been set.  When the flag is set (meaning LSP has changed),
+        // we reset the flag to false and exit out of the loop allowing the request to close.
+        // The client will automatically trigger a new request as soon as we close it, bringing us up to date on diagnostics.
+        while (Interlocked.CompareExchange(ref _lspChanged, value: 0, comparand: 1) == 0)
+        {
+            // There have been no changes between now and when the last request finished - we will hold the connection open while we poll for changes.
+            await Task.Delay(TimeSpan.FromMilliseconds(100), cancellationToken).ConfigureAwait(false);
+        }
+
+        context.TraceInformation("Closing workspace/diagnostics request");
+        // We've hit a change, so we close the current request to allow the client to open a new one.
+        return;
+    }
+
+    public void Dispose()
+    {
+        _workspaceManager.LspTextChanged -= OnLspTextChanged;
+        _workspaceRegistrationService.LspSolutionChanged -= OnLspSolutionChanged;
+    }
```

Comment:
can you please put Dispose right after constructor?  i find it much easier to reason about that way.  :)